### ISSUE TITLE
Guidance to replace A.Lang.sub() function?

### ISSUE THREAD

**Issue by ambrinchaudhary** [liferay/liferay-frontend-guidelines#34]
What should we use to replace A.Lang.sub() function?

**Comment by ambrinchaudhary** [comment]
Example of use: 
var uri = A.Lang.sub(this.portletURL, {
					selectedCategories: this.selectedItems
						.map(item => item.value)
						.join(),
					singleSelect: this.singleSelect,
					vocabularyIds: this.vocabularyIds.concat()
				});

Where url is like http://localhost:8080/group/guest/~/control_panel/manage?...&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_selectedCategories=**{selectedCategories}**&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_singleSelect=**{singleSelect}**&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_vocabularyIds=**{vocabularyIds}**&p_p_auth=i6ntidrM

**Comment by boton** [comment]
Maybe something naive like that:

```js
const sub = (str, obj) => str.replace(/\{([^}]+)\}/g, (_, m) => obj[m]);
```

```js
console.log(sub(
'_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_selectedCategories={selectedCategories}&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_singleSelect={singleSelect}&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_vocabularyIds={vocabularyIds}&p_p_auth=i6ntidrM',
    {selectedCategories: 'hola', singleSelect: 'ke', vocabularyIds: 'ase'}
))

// prints: com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_selectedCategories=hola&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_singleSelect=ke&_com_liferay_asset_categories_selector_web_portlet_AssetCategoriesSelectorPortlet_vocabularyIds=ase&p_p_auth=i6ntidrM
```

**Comment by ambrinchaudhary** [comment]
Thank you @boton. Should we put this in portal, or in metal?

**Comment by julien** [comment]
@boton , @ambrinchaudhary, this might seem like a silly question but couldn't we use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)?

**Comment by jbalsas** [comment]
Hey @julien, they already have the template as a string, so I'm not sure there's a way to turn that into a template literal without eval'ing or some other type of evil 😂 

If I remember correctly, YUI's `Lang.sub` supported nested access, like `a.b.c`, so a simple replacement to what @boton suggest might trigger some edge case regressions...

The implementation in any case should be simple enough, so maybe we could just port over YUI's [`Lang.sub`](https://github.com/yui/yui3/blob/master/src/yui/js/yui-lang.js#L251-L304) and its [tests](https://github.com/yui/yui3/blob/master/src/yui/tests/unit/assets/lang-test.js#L183-L246) over to `frontend-js-web` and use it from there...

Maybe there's a better way for this.

@julien, @wincent, @bryceosterhaus, thoughts?

**Comment by wincent** [comment]
Yeah, we could port it. Although I'd be interested in cataloging the use cases a bit. It might be for simple instances we can use a simpler implementation (without nesting) — especially if we see that there aren't any uses of nested access.

The other question though is, is this the right abstraction anyway for building a URL? Is whatever getting substituted appropriately URI-encoded if necessary in some other place?

**Comment by ambrinchaudhary** [comment]
@wincent yes, URI was being encoded properly (I just edited the example). 
I'll check if we are using nested access in portal.

**Comment by jbalsas** [comment]
Hey @wincent, that's certainly a fair point. 

For URL manipulation, @kresimir-coko is working on [Modernize PortletURL utilities](https://issues.liferay.com/browse/LPS-94714), so maybe we could standardize around it.

`Lang.sub` is also used for template-like interpolation:

```javascript
var TPL_SIMPLE_MENU_ITEM = '<li class="{cssClass}" data-id="{id}">{icon} {caption}</li>';

var li = A.Node.create(
	Lang.sub(TPL_SIMPLE_MENU_ITEM, {
		cssClass: cssClass,
		icon: icon,
		id: id
	})
);
```

Some of these could certainly be directly replaced by template literals if we were transpiling that code (which we are not), but in other cases, if the template string is passed around we might still need a `.sub` or `.template` method, à la lodash...

**Comment by wincent** [comment]
`<aside>`I also wonder about the injection-saftey of constructing HTML with unsafe-by-default template mechanisms that use concatenation and interpolation.`</aside>`

**Comment by jbalsas** [comment]
So... if there's no better option... maybe we should just leave those as they are? Based on:
- We will certainly deal with html in a different way in new (React) code
- URL manipulation shouldn't be handled as templates

Should we just leave this as is?

**Comment by wincent** [comment]
Probably a combination of all three things that you said:

For URLs:

> For URL manipulation, @kresimir-coko is working on [Modernize PortletURL utilities](https://issues.liferay.com/browse/LPS-94714)

For templating:

> we could just port over YUI's [`Lang.sub`](https://github.com/yui/yui3/blob/master/src/yui/js/yui-lang.js#L251-L304) and its [tests](https://github.com/yui/yui3/blob/master/src/yui/tests/unit/assets/lang-test.js#L183-L246) over to `frontend-js-web` and use it from there...

And in the future:

> We will certainly deal with html in a different way in new (React) code

**Comment by bryceosterhaus** [comment]
For lang keys specifically I think porting over is probably our best bet. Unsure about the html and url aspect of it. 

For what its worth, this is the util we would use in Faro and it seemed to work pretty well for instances of language keys where we would sub in either strings or jsx elements.

```js
const SPLIT_REGEX = /({\d+})/g;

/**
 * Utility function for substituting variables into language keys.
 *
 * @example
 * sub(Liferay.Language.get('search-x'), ['all'])
 * => 'search all'
 * sub(Liferay.Language.get('search-x'), [<b>all<b>], false)
 * => 'search <b>all</b>'
 *
 * @param {string} langKey This is the language key used from our properties file
 * @param {string} args Arguments to pass into language key
 * @param {string} join Boolean used to indicate whether to call `.join()` on
 * the array before it is returned. Use `false` if subbing in JSX.
 * @return {(string|Array)}
 */
export function sub(langKey, args, join = true) {
	const keyArray = langKey.split(SPLIT_REGEX).filter(val => val.length !== 0);

	for (let i = 0; i < args.length; i++) {
		const arg = args[i];

		const indexKey = `{${i}}`;

		let argIndex = keyArray.indexOf(indexKey);

		while (argIndex >= 0) {
			keyArray.splice(argIndex, 1, arg);

			argIndex = keyArray.indexOf(indexKey);
		}
	}

	return join ? keyArray.join('') : keyArray;
}
```
